Bug 1712072 - Update distribution docs to new structure (#1906)
mozilla/glean: docs/user/reference/metrics/timing_distribution.md

fun onPageStart(e: Event) {
    timerId = Pages.pageLoad.start()
}
```

</div>
<div data-lang="Java" class="tab">

```Java
import mozilla.components.service.glean.GleanTimerId;
import org.mozilla.yourApplication.GleanMetrics.Pages;

GleanTimerId timerId;

void onPageStart(Event e) {
    timerId = Pages.INSTANCE.pageLoad().start();
}
```

</div>
<div data-lang="Swift" class="tab">

```Swift
import Glean

var timerId : GleanTimerId

func onPageStart() {
    timerId = Pages.pageLoad.start()
}
```

</div>
<div data-lang="Python" class="tab">

```Python
from glean import load_metrics
metrics = load_metrics("metrics.yaml")

class PageHandler:
    def __init__(self):
        self.timer_id = None

    def on_page_start(self, event):
        self.timer_id = metrics.pages.page_load.start()
```

</div>
<div data-lang="Rust" class="tab">

```rust
use glean_metrics::pages;

fn on_page_start() {
    self.timer_id = pages::page_load.start();
}
```

</div>

> <div data-lang="JavaScript" class="tab"></div>

Reviewer: ```suggestion
<div data-lang="JavaScript" class="tab" data-bug="1716954"></div>
```

Stops tracking time for the provided metric and associated timer id.

Adds a count to the corresponding bucket in the timing distribution.
This will record an error if `start` was not called.

{{#include ../../../shared/tab_header.md}}

<div data-lang="Kotlin" class="tab">

```Kotlin
import org.mozilla.yourApplication.GleanMetrics.Pages

fun onPageLoaded(e: Event) {
    Pages.pageLoad.stopAndAccumulate(timerId)
}
```

</div>
<div data-lang="Java" class="tab">

```Java
import org.mozilla.yourApplication.GleanMetrics.Pages;

void onPageLoaded(Event e) {
    Pages.INSTANCE.pageLoad().stopAndAccumulate(timerId);
}
```

</div>
<div data-lang="Swift" class="tab">

```Swift
import Glean

func onPageLoaded() {
    Pages.pageLoad.stopAndAccumulate(timerId)
}
```

</div>
<div data-lang="Python" class="tab">

```Python
from glean import load_metrics
metrics = load_metrics("metrics.yaml")

class PageHandler:
    def on_page_loaded(self, event):
        metrics.pages.page_load.stop_and_accumulate(self.timer_id)
```

</div>
<div data-lang="Rust" class="tab">

```Rust
use glean_metrics::pages;

fn on_page_loaded() {
    pages::page_load.stop_and_accumulate(self.timer_id);
}
```

</div>

> <div data-lang="JavaScript" class="tab"></div>

Reviewer: ```suggestion
<div data-lang="JavaScript" class="tab" data-bug="1716954"></div>
```

<div data-lang="Firefox Desktop" class="tab">

**C++**

```c++
#include "mozilla/glean/GleanMetrics.h"

mozilla::glean::pages::page_load.StopAndAccumulate(std::move(timerId));
```

**JavaScript**

```js
Glean.pages.pageLoad.stopAndAccumulate(timerId);
```

</div>

{{#include ../../../shared/tab_footer.md}}

#### Recorded errors

> * `invalid_value`: If recording a negative timespan.

Author: ```suggestion
* [`invalid_value`](../../user/metrics/error-reporting.md): If recording a negative timespan.
```

> * `invalid_state`: If a non-existing/stopped timer is stopped again.

Author: ```suggestion
* [`invalid_state`](../../user/metrics/error-reporting.md): If a non-existing/stopped timer is stopped again.
```

> * `invalid_overflow`: If recording a time longer than the maximum for the given unit.

Reviewer: Linkify as above.

Author: ```suggestion
* [`invalid_overflow`](../../user/metrics/error-reporting.md): If recording a time longer than the maximum for the given unit.
```

### `measure`

For convenience one can measure the time of a function or block of code.

{{#include ../../../shared/tab_header.md}}

<div data-lang="Kotlin" class="tab">

```Kotlin
import org.mozilla.yourApplication.GleanMetrics.Pages

Pages.pageLoad.measure {
    // Load a page
}
```

</div>
<div data-lang="Java" class="tab"></div>
<div data-lang="Swift" class="tab">

```Swift
import Glean

Pages.pageLoad.measure {
    // Load a page
}
```

</div>
<div data-lang="Python" class="tab">

```Python
from glean import load_metrics
metrics = load_metrics("metrics.yaml")

with metrics.pages.page_load.measure():
    # Load a page
```

</div>
<div data-lang="Rust" class="tab"></div>

> <div data-lang="JavaScript" class="tab"></div>

Reviewer: ```suggestion
<div data-lang="JavaScript" class="tab" data-bug="1716954"></div>
```

fun onPageError(e: Event) {
    Pages.pageLoad.cancel(timerId)
}
```

</div>
<div data-lang="Java" class="tab">

```Java
import org.mozilla.yourApplication.GleanMetrics.Pages;

fun onPageError(e: Event) {
    Pages.INSTANCE.pageLoad().cancel(timerId);
}
```

</div>
<div data-lang="Swift" class="tab">

```Swift
import Glean

func onPageError() {
    Pages.pageLoad.cancel(timerId)
}
```

</div>
<div data-lang="Python" class="tab">

```Python
from glean import load_metrics
metrics = load_metrics("metrics.yaml")

class PageHandler:
    def on_page_error(self, event):
        metrics.pages.page_load.cancel(self.timer_id)
```

</div>
<div data-lang="Rust" class="tab">

```Rust
use glean_metrics::pages;

fn on_page_error() {
    pages::page_load.cancel(self.timer_id);
}
```

</div>
<div data-lang="JavaScript" class="tab"></div>
<div data-lang="Firefox Desktop" class="tab">

**C++**

```c++
#include "mozilla/glean/GleanMetrics.h"

mozilla::glean::pages::page_load.Cancel(std::move(timerId));
```

**JavaScript**

```js
Glean.pages.pageLoad.cancel(timerId);
```

</div>

{{#include ../../../shared/tab_footer.md}}

## Testing API

### `testGetValue`

> Gets the recorded value for a given counter metric.

Reviewer: ```suggestion
Gets the recorded value for a given timing distribution metric.
```

// Get snapshot.
DistributionData snapshot = pages.INSTANCE.pageLoad().testGetValue();

// Usually you don't know the exact timing values,
// but how many should have been recorded.
assertEquals(1L, snapshot.getCount());
```

</div>
<div data-lang="Swift" class="tab">

```Swift
@testable import Glean

// Get snapshot.
let snapshot = try! pages.pageLoad.testGetValue()

// Usually you don't know the exact timing values,
// but how many should have been recorded.
XCTAssertEqual(1, snapshot.count)
```

</div>
<div data-lang="Python" class="tab">

```Python
from glean import load_metrics
metrics = load_metrics("metrics.yaml")

# Get snapshot.
snapshot = metrics.pages.page_load.test_get_value()

# Usually you don't know the exact timing values,
# but how many should have been recorded.
assert 1 == snapshot.count
```

</div>
<div data-lang="Rust" class="tab">

```rust
use glean::ErrorType;
use glean_metrics::pages;

// Get snapshot
let snapshot = pages::page_load.test_get_value(None).unwrap();

// Usually you don't know the exact timing values,
// but how many should have been recorded.
assert_eq!(1, snapshot.values.len());
```

</div>

> <div data-lang="JavaScript" class="tab"></div>

Reviewer: ```suggestion
<div data-lang="JavaScript" class="tab" data-bug="1716954"></div>
```

<div data-lang="Firefox Desktop" class="tab">

**C++**

```c++
#include "mozilla/glean/GleanMetrics.h"

// Does it have an expected values?
const data = mozilla::glean::pages::page_load.TestGetValue().value().unwrap();
ASSERT_TRUE(data.sum > 0);
```

**JavaScript**

```js
Assert.ok(Glean.pages.pageLoad.testGetValue().sum > 0);
```

</div>

{{#include ../../../shared/tab_footer.md}}

### `testHasValue`

> Whether or not **any** value was recorded for a given counter metric.

Reviewer: ```suggestion
Whether or not **any** value was recorded for a given timing distribution metric.
```

{{#include ../../../shared/tab_header.md}}

<div data-lang="Kotlin" class="tab">

```Kotlin
import org.mozilla.yourApplication.GleanMetrics.Pages

// Was anything recorded?
assertTrue(Pages.pageLoad.testHasValue())
```

</div>

<div data-lang="Java" class="tab">

```Java
import org.mozilla.yourApplication.GleanMetrics.Pages;

// Was anything recorded?
assertTrue(Pages.INSTANCE.pageLoad().testHasValue());
```

</div>
<div data-lang="Swift" class="tab">

```Swift
@testable import Glean

// Was anything recorded?
XCTAssert(pages.pageLoad.testHasValue())
```

</div>
<div data-lang="Python" class="tab">

```Python
from glean import load_metrics
metrics = load_metrics("metrics.yaml")

# Was anything recorded?
assert metrics.pages.page_load.test_has_value()
```

</div>
<div data-lang="Rust" class="tab"></div>
<div data-lang="JavaScript" class="tab"></div>
<div data-lang="Firefox Desktop" class="tab"></div>

{{#include ../../../shared/tab_footer.md}}

### `testGetNumRecordedErrors`

> Gets number of errors recorded for a given counter metric.

Reviewer: ```suggestion
Gets number of errors recorded for a given timing distribution metric.
```

<div data-lang="Kotlin" class="tab">

```Kotlin
import org.mozilla.yourApplication.GleanMetrics.Pages

// Assert that no errors were recorded.
assertEquals(0, Pages.pageLoad.testGetNumRecordedErrors(ErrorType.InvalidValue))
```

</div>
<div data-lang="Java" class="tab">

```Java
import org.mozilla.yourApplication.GleanMetrics.Pages;

// Assert that no errors were recorded.
assertEquals(
    0,
    Pages.INSTANCE.pageLoad().testGetNumRecordedErrors(
        ErrorType.InvalidValue
    )
);
```

</div>
<div data-lang="Swift" class="tab">

```Swift
@testable import Glean

// Assert that no errors were recorded.
XCTAssertEqual(0, pages.pageLoad.testGetNumRecordedErrors(.invalidValue))
```

</div>
<div data-lang="Python" class="tab">

```Python
from glean import load_metrics
metrics = load_metrics("metrics.yaml")

# Assert that no errors were recorded.
assert 0 == metrics.pages.page_load.test_get_num_recorded_errors(
    ErrorType.INVALID_VALUE
)
```

</div>
<div data-lang="Rust" class="tab">

```Rust
use glean::ErrorType;
use glean_metrics::pages;

assert_eq!(
    0,
    pages::page_load.test_get_num_recorded_errors(ErrorType::InvalidValue)
);
```

</div>

> <div data-lang="JavaScript" class="tab"></div>

Reviewer: ```suggestion
 <div data-lang="JavaScript" class="tab" data-bug="1716954"></div>
```

Author: ```suggestion
<div data-lang="JavaScript" class="tab" data-bug="1716954"></div>
```

<div data-lang="Firefox Desktop" class="tab"></div>

{{#include ../../../shared/tab_footer.md}}

## Metric parameters

> Example custom distribution metric definition:

Reviewer: ```suggestion
Example timing distribution metric definition:
```
